examples/platforms/simulation/trel.c

```diff
@@ -208,83 +227,167 @@ static void sendPendingTxPackets(void)
+    case TREL_DNSSD_REMOVE_SERVICE_MESSAGE:
+        memset(&peerInfo, 0, sizeof(peerInfo));
+        peerInfo.mRemoved   = (aMessage->mType == TREL_DNSSD_REMOVE_SERVICE_MESSAGE);
+        peerInfo.mTxtData   = aMessage->mData;
+        peerInfo.mTxtLength = (uint8_t)(aMessage->mDataLength);
+        peerInfo.mSockAddr  = aMessage->mSockAddr;
+        otPlatTrelHandleDiscoveredPeerInfo(aInstance, &peerInfo);
+        break;
+    }
+
+exit:
+    return;
 }
 
-void otPlatTrelUdp6SubscribeMulticastAddress(otInstance *aInstance, const otIp6Address *aMulticastAddress)
+//---------------------------------------------------------------------------------------------------------------------
+// otPlatTrel
+
+void otPlatTrelEnable(otInstance *aInstance, uint16_t *aUdpPort)
 {
     OT_UNUSED_VARIABLE(aInstance);
 
-    sMulticastAddress = *aMulticastAddress;
+    *aUdpPort = sUdpPort;
 
 #if DEBUG_LOG
-    fprintf(stderr, "\n[trel-udp6] otPlatTrelUdp6SubscribeMulticastAddress(aMulticastAddress:%s)\n",
-            ip6AddrToString(aMulticastAddress));
+    fprintf(stderr, "\r\n[trel] otPlatTrelEnable() *aUdpPort=%u\r\n", *aUdpPort);
 #endif
+
+    if (!sEnabled)
+    {
+        sEnabled = true;
+        sendBrowseMessage();
+    }
 }
 
-otError otPlatTrelUdp6SendTo(otInstance *        aInstance,
-                             const uint8_t *     aBuffer,
-                             uint16_t            aLength,
-                             const otIp6Address *aDestAddress)
+void otPlatTrelDisable(otInstance *aInstance)
 {
-    otError error = OT_ERROR_NONE;
-    Packet *packet;
-
     OT_UNUSED_VARIABLE(aInstance);
 
-    otEXPECT(sEnabled);
-    otEXPECT_ACTION(sNumPendingTx < TREL_MAX_PENDING_TX, error = OT_ERROR_ABORT);
-    otEXPECT_ACTION(aLength <= TREL_MAX_PACKET_SIZE, error = OT_ERROR_ABORT);
+#if DEBUG_LOG
+    fprintf(stderr, "\r\n[trel] otPlatTrelDisable()\r\n");
+#endif
 
-    packet                          = &sPendingTx[sNumPendingTx++];
-    packet->mHeader.mSrcIp6Address  = sUnicastAddress;
-    packet->mHeader.mDestIp6Address = *aDestAddress;
-    packet->mHeader.mPort           = sUdpPort;
-    packet->mLength                 = aLength + sizeof(PacketHeader);
-    memcpy(packet->mPayload, aBuffer, aLength);
+    if (sEnabled)
+    {
+        sEnabled = false;
 
-exit:
-    return error;
+        if (sServiceRegistered)
+        {
+            sendServiceMessage(TREL_DNSSD_REMOVE_SERVICE_MESSAGE);
+            sServiceRegistered = false;
+        }
+    }
 }
 
-otError otPlatTrelUdp6SetTestMode(otInstance *aInstance, bool aEnable)
+void otPlatTrelRegisterService(otInstance *aInstance, uint16_t aPort, const uint8_t *aTxtData, uint8_t aTxtLength)
 {
     OT_UNUSED_VARIABLE(aInstance);
 
-    sEnabled = aEnable;
+    assert(aTxtLength <= TREL_MAX_SERVICE_TXT_DATA_LEN);
 
-    if (!sEnabled)
+    if (sServiceRegistered)
     {
-        sNumPendingTx = 0;
+        sendServiceMessage(TREL_DNSSD_REMOVE_SERVICE_MESSAGE);
```

Comment:
We need to remove a previous registration before replacing with new one, so that the old service is removed on all peers. Note that the receivers do not track who added which service (they track a list of services),  so it is up to original device which registered it to remove it.

Reply:
Will peers remove the old service using `extaddr` or `SockAddr` matches?

Reply:
It uses both.

Reply:
Fine to me to keep it as is. 

I meant that even if `sendServiceMessage(TREL_DNSSD_REMOVE_SERVICE_MESSAGE);` is  not called, the service should still be replaced by the new TREL service, right?

Reply:
Yes. That's correct.